### PR TITLE
Major refactor of TurbulenceClosures module plus more tests for HydrostaticFreeSurfaceModel

### PR DESCRIPTION
**Pull request by glwagner**
This PR implements a major refactor of the `TurbulenceClosures` module. In particular, it "standardizes" how turbulence closures specify diffusive fluxes. Each turbulence closure now defines 12 functions (or opts in to functionality via subtyping, eg `AbstractEddyViscosityClosure`):

* `diffusive_flux_x`, `diffusive_flux_y`, and `diffusive_flux_z` for tracer diffusive fluxes
* `viscous_flux_ux`, `viscous_flux_uy`, and `viscous_flux_uz` for viscous fluxes of x-momentum
* 6 other functions similar to `viscous_flux_ux`, etc for `y` and `z` momentum components

Then there is a single implementation of `∇_κ_∇c`, for example, which calculates the divergence of the diffusive fluxes in a way that generalizes to curvilinear grids.

This refactor was not that difficult because many turbulence closures had functions analogous to these --- they were just called by different names. So in a way this PR simply establishes standard notation.

A major advantage of standard notation is that we can now correctly specify flux boundary conditions even when we cannot fill halos for this purpose. For example, right now we fill halos so that (fingers crossed!) boundary fluxes are zero when the diffusion operator is evaluated on boundary-adjacent cells, and then in a second step add the specific boundary fluxes. But with a function `diffusive_flux_x`, we can now "reverse" boundary fluxes and add specified fluxes generically. This PR therefore makes progress towards #1400.

Work on this PR started with a few extra tests for `HydrostaticFreeSurfaceModel` (though its focus is not quite different). So as a bonus this PR also adds tests for `HydrostaticFreeSurfaceModel` (which may need to be disabled if they fail on the GPU so that we can get this PR in sooner rather than later).

### CODE REVIEW

File: src/Operators/Operators.jl [hunk 1]
```diff
@@ -4,12 +4,16 @@ export
...
     Azᵃᵃᵃ, Azᶠᶠᵃ, Azᶜᶜᵃ, Azᶠᶜᵃ, Azᶜᶠᵃ,
-    Vᵃᵃᶜ, Vᵃᵃᶠ, Vᶜᶜᶜ,
+    Vᵃᵃᶜ, Vᵃᵃᶠ, Vᶜᶜᶜ, Vᶠᶜᶜ, Vᶜᶠᶜ, Vᶜᶜᶠ,
     δxᶜᵃᵃ, δxᶠᵃᵃ, δyᵃᶜᵃ, δyᵃᶠᵃ, δzᵃᵃᶜ, δzᵃᵃᶠ,
-    Ax_ψᵃᵃᶠ, Ax_ψᵃᵃᶜ, Ay_ψᵃᵃᶠ, Ay_ψᵃᵃᶜ, Az_ψᵃᵃᵃ, Ax_uᶠᶜᶜ, Ay_vᶜᶠᶜ, Az_wᶜᶜᵃ,
+    Ax_ψᵃᵃᶠ, Ax_ψᵃᵃᶜ, Ay_ψᵃᵃᶠ, Ay_ψᵃᵃᶜ, Az_ψᵃᵃᵃ,
+    Ax_uᶠᶜᶜ, Ay_vᶜᶠᶜ, Az_wᶜᶜᵃ,
+    Ax_ζᶠᶠᶜ, Ay_ζᶠᶠᶜ, Ax_ηᶠᶜᶠ, Az_ηᶠᶜᵃ, Ay_ξᶜᶠᶠ, Az_ξᶜᶠᵃ,
+    Az_wᶜᶜᵃ, Az_ηᶠᶜᵃ, Az_ξᶜᶠᵃ,
+    Ax_cᶜᶜᶜ, Ay_cᶜᶜᶜ, Az_cᶜᶜᵃ,
```
Comment (ali-ramadhan) [inline src/Operators/Operators.jl:16]: `u`, `v`, `w`, and `c` should be pretty clear but just wondering if there's a pattern to using `ψ` vs. `ζ` vs. `η` vs. `ξ`?

File: src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_minimum_dissipation.jl [hunk 2]
```diff
@@ -173,6 +200,7 @@ Return the diffusive flux divergence `∇ ⋅ (κ ∇ c)` for the turbulence
             + ∂zᵃᵃᶜ(i, j, k, grid, κ_∂z_c, κₑ, c, closure)
            )
 end
+=#
```
Comment (ali-ramadhan) [inline src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_minimum_dissipation.jl:203]: Should this commented out function be nuked?

File: src/TurbulenceClosures/turbulence_closure_implementations/smagorinsky_lilly.jl [hunk 3]
```diff
@@ -106,77 +106,58 @@ end
...
-      ∂xᶜᵃᵃ(i, j, k, grid, κ_∂x_c, closure, c, tracer_index, diffusivities.νₑ)
-    + ∂yᵃᶜᵃ(i, j, k, grid, κ_∂y_c, closure, c, tracer_index, diffusivities.νₑ)
-    + ∂zᵃᵃᶜ(i, j, k, grid, κ_∂z_c, closure, c, tracer_index, diffusivities.νₑ)
-
-)
+@inline ∇_κ_∇c(i, j, k, grid, clock, closure::SmagorinskyLilly, c, ::Val{tracer_index},
+               diffusivities, args...) where tracer_index = (∂xᶜᵃᵃ(i, j, k, grid, κ_∂x_c, closure, c, Val(tracer_index), diffusivities.νₑ) +
+                                                             ∂yᵃᶜᵃ(i, j, k, grid, κ_∂y_c, closure, c, Val(tracer_index), diffusivities.νₑ) +
+                                                             ∂zᵃᵃᶜ(i, j, k, grid, κ_∂z_c, closure, c, Val(tracer_index), diffusivities.νₑ))
+=#
```
Comment (ali-ramadhan) [inline src/TurbulenceClosures/turbulence_closure_implementations/smagorinsky_lilly.jl:160]: Same as above. Nuke?

File: src/TurbulenceClosures/viscous_dissipation_operators.jl [hunk 4]
```diff
@@ -10,80 +32,23 @@
 @inline νᶜᶜᶜ(i, j, k, grid, clock, ν::Function) = ν(xnode(Center, i, grid), ynode(Center, j, grid), znode(Center, k, grid), clock.time)
 @inline νᶠᶠᶜ(i, j, k, grid, clock, ν::Function) = ν(xnode(Face,   i, grid), ynode(Face,   j, grid), znode(Center, k, grid), clock.time)
 @inline νᶠᶜᶠ(i, j, k, grid, clock, ν::Function) = ν(xnode(Face,   i, grid), ynode(Center, j, grid), znode(Face,   k, grid), clock.time)
-@inline νᶜᶠᶠ(i, j, k, grid, clock, ν::Function) = ν(xnode(Face,   i, grid), ynode(Center, j, grid), znode(Face,   k, grid), clock.time)
+@inline νᶜᶠᶠ(i, j, k, grid, clock, ν::Function) = ν(xnode(Center, i, grid), ynode(Face,   j, grid), znode(Face,   k, grid), clock.time)
```
Comment (ali-ramadhan) [inline src/TurbulenceClosures/viscous_dissipation_operators.jl:35]: 👀

File: test/test_hydrostatic_free_surface_models.jl [hunk 5]
```diff
@@ -97,26 +104,70 @@ end
...
+
+            @testset "Time-stepping Rectilinear HydrostaticFreeSurfaceModels [$arch, $topo]" begin
+                @info "  Testing time-stepping Rectilinear HydrostaticFreeSurfaceModels [$arch, $topo]..."
+                @test time_step_hydrostatic_model_works(arch, grid)
+            end
+        end
+
+        rectilinear_grid = RegularRectilinearGrid(size=(1, 1, 1), extent=(1, 1, 1), halo=(3, 3, 3))
+        lat_lon_sector_grid = RegularLatitudeLongitudeGrid(size=(1, 1, 1), longitude=(0, 60), latitude=(15, 75), z=(-1, 0))
+        lat_lon_strip_grid = RegularLatitudeLongitudeGrid(size=(1, 1, 1), longitude=(-180, 180), latitude=(15, 75), z=(-1, 0))
```
Comment (ali-ramadhan) [inline test/test_hydrostatic_free_surface_models.jl:117]: Free energy balance climate model 🌐

**Review by ali-ramadhan — APPROVED** [review]
This is great! I'd be okay with merging without some of the GPU tests (`@test_skip`?).

Looks like at least some of the GPU tests fail because of 

```
Reason: unsupported call to the Julia runtime (call to jl_f_tuple)
```

as part of `compute_w_from_continuity!` : https://buildkite.com/clima/oceananigans/builds/1497#7c3f4596-4a81-4ef8-9e08-40a44f163a86/35-289

Maybe `compute_w_from_continuity!` is missing a `datatuple` or something is missing from an adapted struct?

File: src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_minimum_dissipation.jl (hunk 2 above)
Comment (glwagner) [reply src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_minimum_dissipation.jl:203]: yes! thank you for catching!

File: src/Operators/Operators.jl (hunk 1 above)
Comment (glwagner) [reply src/Operators/Operators.jl:16]: I'm defining the components of vorticity as `(ξ, η, ζ)` (eg the Greek x, y, z for the x, y, z components). I was thinking that adopting specific notation might help us remember and talk about these locations, and also could help avoid bugs? What do you think?

File: src/TurbulenceClosures/turbulence_closure_implementations/smagorinsky_lilly.jl (hunk 3 above)
Comment (glwagner) [reply src/TurbulenceClosures/turbulence_closure_implementations/smagorinsky_lilly.jl:160]: yes!

File: src/TurbulenceClosures/viscous_dissipation_operators.jl (hunk 4 above)
Comment (glwagner) [reply src/TurbulenceClosures/viscous_dissipation_operators.jl:35]: #untestedfeatures

**Comment by glwagner** [comment]
> Maybe compute_w_from_continuity! is missing a datatuple or something is missing from an adapted struct?

But it's only with the `RegularLatitudeLongitudeGrid` -- a model with `RegularRectilinearGrid` passes. 🤔

**Comment by glwagner** [comment]
OHHHH

Might need https://github.com/JuliaGPU/KernelAbstractions.jl/pull/220.

**Comment by glwagner** [comment]
The lack of translation to `CUDA.sind` and `CUDA.cosd` IS indeed the problem. We can hack our way around it for now until `KernelAbstractions` bends to our whims.

File: test/test_hydrostatic_free_surface_models.jl (hunk 5 above)
Comment (glwagner) [reply test/test_hydrostatic_free_surface_models.jl:117]: ahahah it is a bit confusing to think about this "grid point"

File: src/Operators/Operators.jl (hunk 1 above)
Comment (ali-ramadhan) [reply src/Operators/Operators.jl:16]: Ah that makes sense! I think it's a good idea but might be good to add a comment somewhere explaining the choice of function names? What about `ψ`?

**Pull request merged**